### ISSUE TITLE
How to add unsupported language? (`nob`)

### ISSUE THREAD

**Issue by thomasht86** [ylacombe/finetune-hf-vits#18]
Strangely enough, I can see from [MMS coverage](https://dl.fbaipublicfiles.com/mms/misc/language_coverage_mms.html)
that `nob` (Norwegian) is not supported for TTS. 
![image](https://github.com/ylacombe/finetune-hf-vits/assets/24563696/afc11f16-328e-4299-919a-ffe0cf904df4)

What must be done in order to support it?

**Comment by ylacombe** [comment]
Hi, unfortunately I don't think the authors are planning to release other MMS models. 
From our side, to release such a model, we'd need a good Norwegian TTS datasets, do you have such a dataset in mind?
Would you also be interested in training such a model from scratch?
If so, let me know, I can give you some pointers and some help,
Best

**Comment by thomasht86** [comment]
Hi! 
Thanks for the reply! 
There are at least two large datasets available in Norwegian: 
1. https://huggingface.co/datasets/NbAiLab/NPSC
2. https://huggingface.co/datasets/NbAiLab/NST

I have tried running your scripts, and converting a checkpoint from https://huggingface.co/facebook/mms-tts-swe, with the premise that Swedish and Norwegian is quite similar. 
Had to modify [vocab.json](https://huggingface.co/thomasht86/mms-tts-nob/blob/main/vocab.json) manually, by adding two characters `[æ,å]` to map to same token_id as their Swedish "counterparts". 

I have played around with different learning rates, and parameters, but I consistently get infinity for KL loss, and NaN loss after 100 steps or so....

![image](https://github.com/ylacombe/finetune-hf-vits/assets/24563696/9f4142ed-e7f7-4527-b4e8-c40bb0280c3d)

If you could give pointers for training a model from scratch, I could give it a shot. 😊

**Comment by ylacombe** [comment]
How did you initialize the model ? This might have an important role.
[EDIT:] looking at this [model](https://huggingface.co/thomasht86/mms-tts-nob), it seems okay, did you initialize it from scratch ?

Also which hyper-parameters did you use ? I'll recommend using the default one from the Vits original training.

**Comment by thomasht86** [comment]
No, I generated that one from the swedish model, with `convert_original_discriminator_checkpoint.py`. 
But starting with a model initialized from another language might probably require some tricks to finetune..? 

I used the hyperparameters you provided in https://github.com/ylacombe/finetune-hf-vits/tree/main/training_config_examples as basis, but did a "random manual search" from there. 

Where can I find the default ones from original training?

**Comment by ylacombe** [comment]
In that case, here is a snippet that you can modify to initialize from scratch:

```python
from utils.configuration_vits import VitsConfig
from utils.modeling_vits_training import VitsModelForPreTraining
from utils.feature_extraction_vits import VitsFeatureExtractor
from transformers import AutoTokenizer

NEW_REPO_ID = ...

config = VitsConfig.from_pretrained("thomasht86/mms-tts-nob")
VitsModelForPreTraining(config).push_to_hub(NEW_REPO_ID)

VitsFeatureExtractor.from_pretrained("thomasht86/mms-tts-nob").push_to_hub(NEW_REPO_ID)
AutoTokenizer.from_pretrained("thomasht86/mms-tts-nob").push_to_hub(NEW_REPO_ID)
```

In terms of training, I'd advice:
- focusing on a single speaker per model, as it will facilitate training and be of better quality
- follow the original hyper-parameters (learning rate and loss weights): [here](https://github.com/jaywalnut310/vits/blob/2e561ba58618d021b5b8323d3765880f7e0ecfdb/configs/ljs_base.json#L7-L17)

**Comment by JackismyShephard** [comment]
I am in sort of the same situation but looking to finetune MMS for danish (which is very similar to norwegian). 

I am having trouble understanding where the above code snippet fits into the training pipeline.  Should it be executed after converting a checkpoint using the `convert_original_discriminator_checkpoint.py` script?

**Comment by MendoLeo** [comment]
when training from scratch on new dataset, what about tokenizer type, vocab and text_clenear?
